### ISSUE TITLE
[Bug] KafkaRebalance resource created before CC is enabled is never accepted

### ISSUE THREAD

**Issue by scholzj** [strimzi/strimzi-kafka-operator#3636]
When a `KafkaRebalance` resource is created while Cruise Control is not enabled in `Kafka` CR, CO throws a warning that CC needs to be enabled and this is also set in the `KafkaRebalance` status. However, after CC is enabled in `Kafka` CR and deployed, it should pick up the `KafkaRebalance` resource in periodical reconciliation and use it for `KafkaRebalance`. This does not seem to happen currently.

**Comment by ppatierno** [comment]
Yeah what I see is that it's reconciled but in a wrong way. The `KafkaRebalance` operator reconcile the resource but doesn't take any meaningful action and leaves the last NotReady state as follow.

```shell
2020-10-08 10:02:12 INFO  AbstractOperator:217 - Reconciliation #41(timer) KafkaRebalance(default/my-rebalance): KafkaRebalance my-rebalance will be checked for creation or modification
2020-10-08 10:02:12 INFO  KafkaRebalanceAssemblyOperator:319 - Reconciliation #41(timer) KafkaRebalance(default/my-rebalance): Rebalance action from state [NotReady]
2020-10-08 10:02:12 INFO  KafkaRebalanceAssemblyOperator:333 - Reconciliation #41(timer) KafkaRebalance(default/my-rebalance): State updated to [NotReady] with annotation strimzi.io/rebalance=null 
2020-10-08 10:02:12 INFO  AbstractOperator:455 - Reconciliation #41(timer) KafkaRebalance(default/my-rebalance): reconciled
```

I will look into this.

**Comment by ppatierno** [comment]
So actually, it's not a bug but how it was designed. 
I mean, when the resource it's in `NotReady` state, the user has to annotate with `strimzi.io/rebalance=refresh` for pushing the operator to reconsider the rebalance resource once the error has been fixed.
I guess we did it on purpose for not waiting for a reconcile cycle when the error is not in the resource itself but kind of external like in your case the CC not deployed at the beginning.
Without the annotation, the rebalance resource would be reconciled immediately only if there is an error on it and the user updates the resource itself triggering the watch; it could immediately move from `NotReady` to `New` to be reprocessed.
But when the error is kind of external, the only way is waiting for the next reconcile cycle or doing a change on the resource (which doesn't really need), in this case with an annotation.
I see this as the only reason for the need for the "refresh" annotation otherwise we can just say the user to wait for the next reconcile.

@tomncooper any other thoughts on the "refresh" annotation we decided to add for escaping the `NotReady` state?

**Comment by scholzj** [comment]
If it was designed like this, I think this is very bad UX because it works completely differently then any other resource we have. All of them can be created up front and will be picked-up later.

**Comment by tomncooper** [comment]
Well this issue is that we don't want to keep pinging the CC server with mal-formed/mis-configured requests, so we make the user actively tell use when they think the request is correct. The reason the UX for this is different from all the other resources is the Cruise Control (and cluster rebalancing) is different to all the other resources.

**Comment by scholzj** [comment]
I don't think that makes sense. When the CC is already running and I create the resource, I do not need to do any annotation. So why can't it pick the resource which already exists when CC is enabled and deployed?

**Comment by ppatierno** [comment]
I am kind of, in the middle ....
On one side I see the value of the annotation to avoid the operator pinging the CC every 2 minutes when for example the error is in the KafkaRebalance resource and the user doesn't fix it.
On the other side I see the difference with other things and we base everything on reconcile loop.

**Comment by tomncooper** [comment]
So, I have no strong objection if you want to remove the annotation. 

However, whilst the KR could fail for a lot of reasons, most of which will be syntax issues that will not effect CC. The one issue that could come up, is if a user specifies a KR with custom goals (where a relevant OP is not cached). That could involve a lot of work to find an OP on the CC side. If those goals can't be satisfied and CC never finds an OP, then the request will time out eventually (blocking the whole CO reconcile in the process) and move the KR to `NotReady`. Without the annotation it will keep repeating this process until the user notices and stops or refreshes the KR configs. This could significantly extend reconciles for the whole CO.

How likely that is, I can't say. But it is possible.

**Comment by scholzj** [comment]
TBH I do not think when the rebalance fails for any reason it needs to be re-tried again and again. I think the user experience is particularly bad when the resource is created before CC is enabled and is ignored. So it is this particular scenario which I think is troubling here. Not any CC specific errors.

**Comment by ppatierno** [comment]
I see another use case ...
The `KafkaRebalance` resource has to have a label `strimzi.io/cluster: my-cluster` specifying the Kafka cluster it belongs to.
If you do the same steps so ... 1 Deploy Kafka without CC 2. KR without cluster label goes into NotReady 3. Update Kafka with CC ... nothing happen again until you apply the "refresh".
We can handle these two specific use case in a specific way, which means another "state" in the KafkaRebalance operator state machine if we want to leave the "refresh" annotation in all the other cases (i.e. user errors/mistakes in the KafkaRebalance resource).
Is it really worth handling specific cases compared to just remove the "refresh" annotation?

**Comment by ppatierno** [comment]
@scholzj @tomncooper so where we want to head with this? Just handling the above two specific use cases in a different way? I mean no need to "refresh" the `KafkaRebalance` when they happen while still need the annotation for ALL the other cases (that anyway could be more CC related)?

**Comment by scholzj** [comment]
TBH, I think it would be nice to have at least in my case nicer behaviour. But I do not think this has priority since I guess I was so far the only one who complained.

**Comment by ppatierno** [comment]
For now, I will add more details on possible problems in this section https://strimzi.io/docs/operators/master/using.html#proc-fixing-problems-with-kafkarebalance-str where it's described the usage of "refresh" annotation.

**Comment by scholzj** [comment]
Discussed on the Community meeting (22nd of October):
* Annotation should not be needed when the resource is created before CC is deployed or when the cluster label was wrong
* Errors in proposals / rebalance request should still require the annotation

**Comment by scholzj** [comment]
Triaged on 2.8.2022: Should be fixed in #6561
